packages/react-jss/src/createUseStyles.js

```diff
@@ -42,111 +35,75 @@ const createUseStyles = <Theme: {}>(styles: Styles<Theme>, options?: HookOptions
     const context = React.useContext(JssContext)
     const theme = useTheme()
 
-    const [state, dispatch] = React.useReducer(reducer, null, () => {
-      const sheet = createStyleSheet({
-        context,
-        styles,
-        name,
-        theme,
-        index,
-        sheetOptions
-      })
-
-      let dynamicRules
-      let classes
-      if (sheet) {
-        if (context.registry) {
-          context.registry.add(sheet)
-        }
-        dynamicRules = addDynamicRules(sheet, data)
-        classes = getSheetClasses(sheet, dynamicRules)
-      }
-
-      return {
-        sheet,
-        dynamicRules,
-        classes: classes || {}
-      }
-    })
-
-    useEffectOrLayoutEffect(
+    const [sheet, dynamicRules] = React.useMemo(
       () => {
-        if (state.sheet) {
+        const newSheet = createStyleSheet({
+          context,
+          styles,
+          name,
+          theme,
+          index,
+          sheetOptions
+        })
+
+        const newDynamicRules = newSheet ? addDynamicRules(newSheet, data) : null
+
+        if (newSheet) {
           manageSheet({
             index,
             context,
-            sheet: state.sheet,
+            sheet: newSheet,
             theme
           })
         }
 
-        return () => {
-          const {sheet, dynamicRules} = state
-
-          if (!sheet) return
-
-          unmanageSheet({
-            index,
-            context,
-            sheet,
-            theme
-          })
-
-          if (dynamicRules) {
-            removeDynamicRules(sheet, dynamicRules)
-          }
-        }
+        return [newSheet, newDynamicRules]
       },
-      [state.sheet]
+      [context, theme]
     )
 
     useEffectOrLayoutEffect(
       () => {
         // We only need to update the rules on a subsequent update and not in the first mount
-        if (state.sheet && state.dynamicRules && !isFirstMount.current) {
-          updateDynamicRules(data, state.sheet, state.dynamicRules)
+        if (sheet && dynamicRules && !isFirstMount.current) {
+          updateDynamicRules(data, sheet, dynamicRules)
         }
       },
       [data]
     )
 
     useEffectOrLayoutEffect(
```

Comment:
Can't we just move this above the `updateDynamicRules` hook?

Reply:
That cleanup depends on the sheet argument, the one above on data.

Reply:
Seems correct to me